### PR TITLE
[MRG] Support sample weights on everything that uses the SGD loss functions.

### PR DESCRIPTION
**Pull request by fabianp**
I think this might be a killer feature :-)

To make it efficient (i.e. not make the unweighted version slower) and generic to all algorithms I create custom loss functions that are weighted and pass an extra argument (the index i) to the loss function in the algorithm iterations.

TODO:

  - [x] more tests
  - [x] document
  - [x] example

### CODE REVIEW

**Comment by mblondel** [comment]
@pprett for sure would be interested :)

**Comment by pprett** [comment]
very interested indeed :)

**Comment by fabianp** [comment]
Pushed a big update. I've changed a bit the logic and now implement the weighted loss by a wrapper around the loss function so that all loss function are supported with a single new loss (WeightedLoss), instead of defining a new weighted loss for each unweighted loss function. Also added support for {SGD, SVRG, SAGA}, tests and an example (shamelessly taken from @pprett's scikit-learn one). 

Changing the title to [MRG]

**Comment by mblondel** [comment]
That's great, thanks a lot. I personally don't have time to make a thorough review but maybe @pprett or @vene can help :)

**Comment by fabianp** [comment]
@casotto, could you take a look at this?  It is currently blocked by lack of reviews. Since you are a potential user of this feature it would be great to have your opinion

**Comment by mblondel** [comment]
@fabianp If I am not mistaken, you also need to update the utility function which computes Lipschitz constants.

File: lightning/impl/sgd_fast.pyx [hunk 1]
```diff
@@ -53,6 +53,21 @@ cdef class ModifiedHuber(LossFunction):
...
+
+    def __init__(self, np.ndarray[double, ndim=1] sample_weights, LossFunction loss_func):
+        self.sample_weights = sample_weights
+        self.loss_func = loss_func
+
+    cpdef double loss(self, double p, double y, int i):
+        return self.loss_func.loss(p, y, i) * self.sample_weights[i]
+
+    cpdef double get_update(self, double p, double y, int i):
+        return self.loss_func.get_update(p, y, i) * self.sample_weights[i]
```
Comment (mblondel) [inline lightning/impl/sgd_fast.pyx]: Instead of using `WeightedLoss`, have you tried added `sample_weights` directly to `Loss`? I know that this will cause an addition multiplication per sample but I would be curious to know the performance hit.

File: lightning/impl/sgd_fast.pyx [hunk 2]
```diff
@@ -53,6 +53,21 @@ cdef class ModifiedHuber(LossFunction):
...
+cdef class WeightedLoss(LossFunction):
+
+    cdef double[:] sample_weights
+    cdef LossFunction loss_func
+
+    def __init__(self, np.ndarray[double, ndim=1] sample_weights, LossFunction loss_func):
+        self.sample_weights = sample_weights
+        self.loss_func = loss_func
+
+    cpdef double loss(self, double p, double y, int i):
```
Comment (mblondel) [inline lightning/impl/sgd_fast.pyx]: An alternative would be `loss(self, double p, double y, double sw)`. This would require passing sample_weights to each algorithm but the method signature is cleaner. I liked the fact that the loss functions are independent of the dataset.

**Comment by fabianp** [comment]
Good point on the Lipchitz constant @mblondel

File: lightning/impl/sgd_fast.pyx (hunk 2 above)
Comment (fabianp) [inline lightning/impl/sgd_fast.pyx]: If we are willing to pay the extra multiplication at each iteration we might as well replace loss.get_update(...) with sw * loss.get_update(...) inside the algorithm itself (sgd, sag, svrg) and not modify the loss API at all.

Comment (mblondel) [inline lightning/impl/sgd_fast.pyx]: I would say it depends on the performance hit. If it's something like 30%, your design is worth it. If it's more like 5%, I think I would prefer the simpler design.

**Comment by fabianp** [comment]
closing this in favour of #73
